editoast/src/views/timetable/train_schedule.rs

```diff
@@ -590,10 +608,342 @@ async fn occupancy_blocks(
+                    .iter()
+                    .find(|to| *to == &track_offset1))
+                .expect("Invalid track offset")
+                .track
+                .clone()
+        }
+    };
+
+    Some((
+        track_section.to_string(),
+        TrackOccupancy {
+            time_begin: train_schedule.start_time + Duration::milliseconds(time as i64),
+            duration: stop_duration,
+            train_schedule_id: train_schedule.id,
+        },
+    ))
+}
+
+fn compute_track_occupancy_without_path_item_id(
+    train_schedule: &models::TrainSchedule,
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
+) -> Option<(String, TrackOccupancy)> {
+    // Iterate over the track offsets linked to the operational point
+    // until we find its corresponding position on the pathfinding route.
+    // Then, retrieve the associated time (via interpolation) and position on the path.
+    for track_offset in operational_point_track_offsets {
+        if let Some(position) = path_projection.get_position(track_offset) {
+            let index = find_index_upper(&report_train.positions, position);
+            let time = interpolate(
+                report_train.positions[index - 1],
+                report_train.positions[index],
+                report_train.times[index - 1],
+                report_train.times[index],
+                position,
+            );
+            return Some((
+                track_offset.track.to_string(),
+                TrackOccupancy {
+                    time_begin: train_schedule.start_time + Duration::milliseconds(time as i64),
+                    duration: PositiveDuration::default(),
+                    train_schedule_id: train_schedule.id,
+                },
+            ));
+        }
+    }
+    None
+}
+
+/// Computes the occupancy of track sections for a given train based on its schedule and path.
+/// If a path item ID is provided, it uses schedule data to determine stop duration and location.
+/// If not, it infers the position and time based on track offsets and interpolation.
+fn compute_track_occupancy(
+    path_item_id: Option<&str>,
+    path_item_positions: &[u64],
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
+    train_schedule: &models::TrainSchedule,
+) -> Option<(String, TrackOccupancy)> {
+    if let Some(path_item_id) = path_item_id {
+        return compute_track_occupancy_with_path_item_id(
+            train_schedule,
+            path_item_id,
+            path_item_positions,
+            path_projection,
+            operational_point_track_offsets,
+            report_train,
+        );
+    };
+
+    compute_track_occupancy_without_path_item_id(
+        train_schedule,
+        path_projection,
+        operational_point_track_offsets,
+        report_train,
+    )
+}
```

Comment:
nitpick: Using `if { return foo; } bar;` is not completely straightforward to express "I'll do this XOR that". How about using `if { foo } else { bar }` instead?